axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/PurchaseOrderLineServiceSupplychainImpl.java

```diff
@@ -202,4 +209,146 @@ public void computeBudgetDistributionSumAmount(
+      Company company,
+      AnalyticAccount analyticAccount)
+      throws AxelorException {
+    AnalyticMoveLine analyticMoveLine =
+        analyticMoveLineService.computeAnalytic(company, analyticAccount);
+
+    analyticMoveLine.setDate(
+        appBaseService.getTodayDate(
+            purchaseOrderLine.getPurchaseOrder() != null
+                ? purchaseOrderLine.getPurchaseOrder().getCompany()
+                : Optional.ofNullable(AuthUtils.getUser())
+                    .map(User::getActiveCompany)
+                    .orElse(null)));
+    analyticMoveLine.setAmount(purchaseOrderLine.getExTaxTotal());
+    analyticMoveLine.setTypeSelect(AnalyticMoveLineRepository.STATUS_FORECAST_ORDER);
+
+    return analyticMoveLine;
+  }
+
+  @Override
+  public PurchaseOrderLine printAnalyticAccount(PurchaseOrderLine line, Company company)
+      throws AxelorException {
+    if (line.getAnalyticMoveLineList() != null
+        && !line.getAnalyticMoveLineList().isEmpty()
+        && company != null) {
+      List<AnalyticMoveLine> analyticMoveLineList = Lists.newArrayList();
+      for (AnalyticAxisByCompany analyticAxisByCompany :
+          accountConfigService.getAccountConfig(company).getAnalyticAxisByCompanyList()) {
+        for (AnalyticMoveLine analyticMoveLine : line.getAnalyticMoveLineList()) {
+          if (analyticMoveLine.getAnalyticAxis().equals(analyticAxisByCompany.getAnalyticAxis())) {
+            analyticMoveLineList.add(analyticMoveLine);
+          }
+        }
+
+        if (!analyticMoveLineList.isEmpty()) {
+
+          AnalyticMoveLine analyticMoveLine = analyticMoveLineList.get(0);
+          if (analyticMoveLineList.size() == 1
+              && analyticMoveLine.getPercentage().compareTo(new BigDecimal(100)) == 0) {
+            AnalyticAccount analyticAccount = analyticMoveLine.getAnalyticAccount();
+            switch (analyticAxisByCompany.getSequence()) {
+              case 0:
+                line.setAxis1AnalyticAccount(analyticAccount);
+                break;
+              case 1:
+                line.setAxis2AnalyticAccount(analyticAccount);
+                break;
+              case 2:
+                line.setAxis3AnalyticAccount(analyticAccount);
+                break;
+              case 3:
+                line.setAxis4AnalyticAccount(analyticAccount);
+                break;
+              case 4:
+                line.setAxis5AnalyticAccount(analyticAccount);
+                break;
+              default:
+                break;
+            }
+          }
+        }
+        analyticMoveLineList.clear();
+      }
+    }
+    return line;
+  }
+
+  public void clearAnalyticInLine(PurchaseOrderLine purchaseOrderLine) {
+    purchaseOrderLine.setAxis1AnalyticAccount(null);
+    purchaseOrderLine.setAxis2AnalyticAccount(null);
+    purchaseOrderLine.setAxis3AnalyticAccount(null);
+    purchaseOrderLine.setAxis4AnalyticAccount(null);
+    purchaseOrderLine.setAxis5AnalyticAccount(null);
+  }
```

Comment:
everything in this service looks like duplicated code. Please search how can we avoid this, maybe a good solution here would be to use an interface on the PurchaseOrderLine and ContractLine models with a dedicated service for this interface.